### PR TITLE
Redirects API - Client support

### PR DESCRIPTION
**Pull request by amanjeev**
Note: This will only work with the dev server right now. That is primarily the reason this is a draft PR.

# Testing instructions

## Project manifest

Create a project manifest with content:

```toml
manifest-version = 1

[products.ferrocene]
release = "@stable-24.05/latest"
packages = [
    "rustc-x86_64-unknown-linux-gnu"
]
```

Note the `release = "@stable-24.05/latest"` line that tells the client to ask the server for the latest release in the stable-24.05 channel.

Save this file. We will call this file `criticalup.toml`.

## Run `install` command

```sh
cargo run -q --package criticalup-dev install --project criticalup.toml
```

The part `--package criticalup-dev` tells cargo to use `criticalup-dev` binary instead of default because we can only test this with dev server as of right now.

## Output

You should see a successful output of 

```sh
 INFO Installing product 'ferrocene' (@stable-24.05/latest)
 INFO Resolved '@stable-24.05/latest' to release 'stable-24.05.0'
 INFO Downloading component 'rustc-x86_64-unknown-linux-gnu' for 'ferrocene' (stable-24.05.0)
 INFO Installing component 'rustc-x86_64-unknown-linux-gnu' for 'ferrocene' (stable-24.05.0)
```

The `@stable-24.05/latest` was resolved to `stable-24.05.0`.

To further check if it is indeed the latest one, you can run 

```sh
cargo run -q --package criticalup-dev run --project criticalup.toml rustc --version
```

This should print

```sh
rustc 1.76.0 (5f6ca3d45 2024-05-22) (Ferrocene by Ferrous Systems)
```

### CODE REVIEW

File: crates/criticalup-core/src/download_server_client.rs [hunk 1]
```diff
@@ -71,9 +72,23 @@ impl DownloadServerClient {
         product: &str,
         release: &str,
     ) -> Result<ReleaseManifest, Error> {
-        let p = format!("/v1/releases/{product}/{release}");
+        // TODO: tests
```
Comment (Hoverbear) [inline crates/criticalup-core/src/download_server_client.rs:75]: Is this blocking this PR or for the future?

File: crates/criticalup-cli/src/commands/install.rs [hunk 2]
```diff
@@ -111,6 +111,12 @@ async fn install_product_afresh(
         .await?;
     let verified_release_manifest = release_manifest_from_server.signed.into_verified(&keys)?;
 
+    if release.starts_with('@') {
```
Comment (Hoverbear) [inline crates/criticalup-cli/src/commands/install.rs:114]: We have a couple places we check for this `@` symbol, I wonder if we should teach the `ProjectManifestProduct` how to know if it's an alias or not to isolate concept. Something like a `fn release_alias(&self) -> Option<String>` which returns the alias if any. (And `release()` maybe needs to be rethought too, as if it's an alias it's not actually a release, is it?)

File: docs/src/criticalup_toml.rst [hunk 3]
```diff
@@ -87,6 +87,18 @@ channels page <https://releases.ferrocene.dev/ferrocene/index.html>`_.
     release = "stable-24.05.0"
     # ...
 
+**Getting the latest release in a channel**
+
+Sometimes you may want the latest release for a given channel. You can use
+the following pattern to achieve that. Here you will get latest release in
+the nightly channel.
```
Comment (Hoverbear) [inline docs/src/criticalup_toml.rst:94]: Let's not recommend people use nightly, let's use stable.

File: docs/src/criticalup_toml.rst [hunk 4]
```diff
@@ -87,6 +87,18 @@ channels page <https://releases.ferrocene.dev/ferrocene/index.html>`_.
     release = "stable-24.05.0"
     # ...
 
+**Getting the latest release in a channel**
+
+Sometimes you may want the latest release for a given channel. You can use
```
Comment (Hoverbear) [inline docs/src/criticalup_toml.rst:92]: This section doesn't match the tone or perspective of other entries, can we align it?

File: crates/criticalup-core/src/download_server_client.rs (hunk 1 above)
Comment (amanjeev) [reply crates/criticalup-core/src/download_server_client.rs]: For the future. I did add one measly test but more can be done once I consolidate the test utils.

File: docs/src/criticalup_toml.rst (hunk 3 above)
Comment (amanjeev) [reply docs/src/criticalup_toml.rst]: Thank you!

File: docs/src/criticalup_toml.rst (hunk 4 above)
Comment (amanjeev) [reply docs/src/criticalup_toml.rst]: I tried to match tone.

File: crates/criticalup-core/src/errors.rs [hunk 5]
```diff
@@ -95,6 +95,14 @@ pub enum Error {
 
     #[error("Failed to load keys into keychain.")]
     KeychainLoadingFailed(#[source] criticaltrust::Error),
+
+    #[error(
+        "The `release` in your project manifest does not seem to be correct.\n  \
+    If you want to specify latest release from a channel, please use '@' symbol as the first \
+    character, then the channel name, then the symbol '/', and then the word 'latest'.\n  \
```
[the remark this answers is not in the document]
Comment (amanjeev) [reply crates/criticalup-core/src/errors.rs]: I made it better and also had a bug that is fixed.

**Comment by amanjeev** [comment]
This work is not being done. Closing.

**Pull request closed without merging**